File: data/src/main/java/com/everis/data/repository/DatabaseDataRepository.java
package com.everis.data.repository;

import com.everis.data.model.IntroMessageEntity;
import com.everis.data.model.IntroMessageEntityDataMapper;
import com.everis.data.network.firebase.FirebaseDBImpl;
import com.everis.domain.model.IntroMessage;
import com.everis.domain.model.LocalContact;
import com.everis.domain.model.P2PUser;
import com.everis.domain.repository.DatabaseRepository;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;

import java.util.HashMap;
import java.util.List;

import rx.Observable;
import rx.Subscriber;
import rx.functions.Func1;

/**
 * Created by everis on 26/04/18.
 */

public class DatabaseDataRepository implements DatabaseRepository{

    private FirebaseDBImpl database;

    public DatabaseDataRepository() {
        this.database = new FirebaseDBImpl();
    }

    @Override
    public Observable<IntroMessage> getIntroMessage() {
        return database.observeValueEvent(FirebaseDatabase.getInstance().getReference().child("intro_message").child("asd123")).map(new Func1<DataSnapshot, IntroMessage>() {
            @Override
            public IntroMessage call(DataSnapshot dataSnapshot) {
                IntroMessageEntity introMessageEntity = dataSnapshot.getValue(IntroMessageEntity.class);
                introMessageEntity.setUid(dataSnapshot.getKey());
                return IntroMessageEntityDataMapper.transform(introMessageEntity);
            }
        });
    }

    @Override
    public Observable<Boolean> getUserEnabled(String userUid) {
        return database.observeValueEvent(FirebaseDatabase.getInstance().getReference().child("structure").child("users").child(userUid)).map(new Func1<DataSnapshot, Boolean>() {
            @Override
            public Boolean call(DataSnapshot dataSnapshot) {
                if(dataSnapshot.exists()) {
                    HashMap<String, Object> userValue = (HashMap<String, Object>) dataSnapshot.getValue();
                    return (Boolean) userValue.get("enable");
                }
                else
                    return false;
            }
        });
    }

    @Override
    public Observable<P2PUser> getUserByPhoneReference(String phoneNumber) {

        return database.observeValueEvent(FirebaseDatabase.getInstance().getReference().child("p2p_users").child(phoneNumber)).map(
                new Func1<DataSnapshot, P2PUser>() {
                    @Override
                    public P2PUser call(DataSnapshot snapshot) {
                        return snapshot.getValue(P2PUser.class);
                    }
                }
        );
    }

    @Override
    public Observable<P2PUser> getUserByPhoneQuery(String phoneNumber) {
        return database.observeValueEvent(FirebaseDatabase.getInstance().getReference().child("p2p_users").orderByChild("phoneNumber").equalTo(phoneNumber)).map(new Func1<DataSnapshot, P2PUser>() {
            @Override
            public P2PUser call(DataSnapshot snapshot) {
                return snapshot.getValue(P2PUser.class);
            }
        });
    }

    @Override
    public Observable<P2PUser> syncUserContacts(List<LocalContact> localContacts, boolean isByQuery) {
        if(isByQuery) {
            return database.observeBatchContactsValueEventListener(localContacts, isByQuery).map(new Func1<DataSnapshot, P2PUser>() {
                @Override
                public P2PUser call(DataSnapshot snapshot) {
                    return snapshot.getChildren().iterator().next().getValue(P2PUser.class);
                }
            });
        }
        else{
            return database.observeBatchContactsValueEventListener(localContacts, isByQuery).map(new Func1<DataSnapshot, P2PUser>() {
                @Override
                public P2PUser call(DataSnapshot snapshot) {
                    if(snapshot.exists())
                        return snapshot.getValue(P2PUser.class);
                    else{
                        P2PUser inactiveUser = new P2PUser();
                        inactiveUser.setPhoneNumber(snapshot.getKey());
                        return inactiveUser;
                    }
                }
            });
        }
    }

    @Override
    public Observable<Void> observeRemoveValue(final String collection, final String innerNode) {
        return Observable.create(new Observable.OnSubscribe<Void>() {
            @Override
            public void call(final Subscriber<? super Void> subscriber) {
                FirebaseDatabase.getInstance().getReference().child(collection).child(innerNode).removeValue(new DatabaseReference.CompletionListener() {
                    @Override
                    public void onComplete(DatabaseError databaseError, DatabaseReference databaseReference) {
                        if(databaseError!=null){
                            subscriber.onError(databaseError.toException());
                        }
                        else{
                            subscriber.onNext(null);
                        }
                    }
                });
            }
        });
    }

}

File: data/src/main/java/com/everis/data/repository/DatabaseDataRepository.kt
package com.everis.data.repository

import com.everis.data.model.IntroMessageEntity
import com.everis.data.model.IntroMessageEntityDataMapper
import com.everis.data.network.firebase.FirebaseDBImpl
import com.everis.domain.model.IntroMessage
import com.everis.domain.model.LocalContact
import com.everis.domain.model.P2PUser
import com.everis.domain.repository.DatabaseRepository

import com.google.firebase.database.FirebaseDatabase

import rx.Observable

class DatabaseDataRepository : DatabaseRepository {

    private val database: FirebaseDBImpl = FirebaseDBImpl()

    override fun getIntroMessage(): Observable<IntroMessage> {
        return database.observeValueEvent(FirebaseDatabase.getInstance().reference.child("intro_message").child("asd123")).map { dataSnapshot ->
            val introMessageEntity = dataSnapshot.getValue(IntroMessageEntity::class.java)
            introMessageEntity!!.uid = dataSnapshot.key
            IntroMessageEntityDataMapper.transform(introMessageEntity)
        }
    }

    override fun getUserEnabled(userUid: String): Observable<Boolean> {
        return database.observeValueEvent(FirebaseDatabase.getInstance().reference.child("structure").child("users").child(userUid)).map { dataSnapshot ->
            if (dataSnapshot.exists()) {
                dataSnapshot.child("enable").value as Boolean
            } else
                false
        }
    }

    override fun getUserByPhoneReference(phoneNumber: String): Observable<P2PUser> {

        return database.observeValueEvent(FirebaseDatabase.getInstance().reference.child("p2p_users").child(phoneNumber)).map { snapshot -> snapshot.getValue(P2PUser::class.java) }
    }

    override fun getUserByPhoneQuery(phoneNumber: String): Observable<P2PUser> {
        return database.observeValueEvent(FirebaseDatabase.getInstance().reference.child("p2p_users").orderByChild("phoneNumber").equalTo(phoneNumber)).map { snapshot -> snapshot.getValue(P2PUser::class.java) }
    }

    override fun observeRemoveValue(collection: String, node: String): Observable<Void> {
        return Observable.create { subscriber ->
            FirebaseDatabase.getInstance().reference.child(collection).child(node).removeValue { databaseError, databaseReference ->
                if (databaseError != null) {
                    subscriber.onError(databaseError.toException())
                } else {
                    subscriber.onNext(null)
                }
            }
        }
    }

    override fun syncUserContacts(localContacts: List<LocalContact>, isByQuery: Boolean): Observable<P2PUser> {
        return if (isByQuery) {
            database.observeBatchContactsValueEventListener(localContacts, isByQuery).map { snapshot -> snapshot.children.iterator().next().getValue(P2PUser::class.java) }
        } else {
            database.observeBatchContactsValueEventListener(localContacts, isByQuery).map { snapshot ->
                if (snapshot.exists())
                    snapshot.getValue(P2PUser::class.java)
                else {
                    val inactiveUser = P2PUser()
                    inactiveUser.phoneNumber = snapshot.key
                    inactiveUser
                }
            }
        }
    }
}
